libbpf-tools/futexctn.bpf.c

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2023 Wenbo Zhang
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+
+#include "futexctn.h"
+#include "bits.bpf.h"
+
+#define MAX_ENTRIES	10240
+
+#define FUTEX_WAIT             0
+#define FUTEX_PRIVATE_FLAG     128
+#define FUTEX_CLOCK_REALTIME   256
+#define FUTEX_CMD_MASK         ~(FUTEX_PRIVATE_FLAG | FUTEX_CLOCK_REALTIME)
+
+const volatile bool targ_per_stack = false;
+const volatile bool targ_ms = false;
+const volatile pid_t targ_pid = 0;
+const volatile pid_t targ_tid = 0;
+
+struct val_t {
+	u64 ts;
+	u32 uaddr;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, struct val_t);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(key_size, sizeof(u32));
+} stackmap SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct hist_key);
+	__type(value, struct hist);
+} hists SEC(".maps");
+
+static struct hist initial_hist;
+
+SEC("tracepoint/syscalls/sys_enter_futex")
+int futex_enter(struct trace_event_raw_sys_enter *ctx)
+{
+	struct val_t v = {};
+	u64 pid_tgid;
+	u32 tid;
+
+	if (((int)ctx->args[1] & FUTEX_CMD_MASK) != FUTEX_WAIT)
+		return 0;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+	tid = (__u32)pid_tgid;
+	if (targ_pid && targ_pid != pid_tgid >> 32)
+		return 0;
+	if (targ_tid && targ_tid != tid)
+		return 0;
+	v.uaddr = ctx->args[0];
```

Comment:
We can't store a pointer in a 32-bit value.

Reply:
The `mutex` is defined as below (in [glibc 2.35](https://github.com/bminor/glibc/blob/4290aed05135ae4c0272006442d147f2155e70d7/sysdeps/x86/nptl/bits/struct_mutex.h#L22)):

```c
struct __pthread_mutex_s
{
  int __lock;
  unsigned int __count;
  int __owner;
#ifdef __x86_64__
  unsigned int __nusers;
...
}
```
So we can see the value of `&__lock` is also the address of `pthread_mutex_t *mutex`. The glibc will pass it to the syscall `futex`: 
```c
futex_wait ((unsigned int *) &mutex->__data.__lock, oldval, ...
```

So It's used as ((unsigned long)(REC->uaddr)).